### ISSUE TITLE
Different memory functions

### ISSUE THREAD

**Issue by NazzarenoMarziale** [babelouest/ulfius#63]
I used o_set_alloc_funcs() with my memory allocation functions. All is well, but when microhttpd server tries to close request rises an error.

/lib64/libc.so.6(+0x7f5e4)[0x7f673227b5e4]                                                                                                                                                             
/lib64/libmicrohttpd.so.10(MHD_destroy_response+0x4a)[0x7f672f02e7ba]                                                                                                                                  
/lib64/libmicrohttpd.so.10(MHD_connection_handle_idle+0x251)[0x7f672f0265f1]                                                                                                                           
/lib64/libmicrohttpd.so.10(+0x92cb)[0x7f672f0292cb]

I think that it uses it's own malloc and not my malloc version.
Maybe inside o_set_alloc_funcs you have to comunicate also to microhttpd server that new memory functions as it is done inside ulfius_send_http_streming_request for CURL global_init_mem.

**Comment by babelouest** [comment]
Hello @NazzarenoMarziale ,

Thanks for the feedback, unfortunately it looks like there is no way to specify memory allocation functions to MHD library. MHD uses very few memory allocation/deallocation and when it dies, it uses directly `malloc`, [for example](https://gnunet.org/git/libmicrohttpd.git/tree/src/microhttpd/daemon.c#n376).

Although, I think I know how to fix the problem, ulfius uses the following code to pass the response body to MHD daemon:
```C
MHD_create_response_from_buffer (response_buffer_len, response_buffer, MHD_RESPMEM_MUST_FREE );
```

Where `MHD_RESPMEM_MUST_FREE` means that MHD will free `response_buffer` by itself. I think a better way to do so is to replace all `MHD_create_response_from_buffer` by:

```C
MHD_create_response_from_data(response_buffer_len, response_buffer, 0, 1);
```

Then, in the function `ulfius_webservice_dispatcher`, right before [return mhd_ret;](https://github.com/babelouest/ulfius/blob/master/src/ulfius.c#L648), insert `o_free(response_buffer);`

I'll provide a patch in the branch 2.4 which will be released soon.
But if you're in a hurry, I can patch the master branch directly.

**Comment by NazzarenoMarziale** [comment]
Hi @babelouest thanks for the response.
I applied you changes and now it works, but the `MHD_create_response_from_data` function is deprecated. 
So i tried to change:
`MHD_create_response_from_buffer (response_buffer_len, response_buffer, MHD_RESPMEM_MUST_FREE );`
with `MHD_RESPMEM_MUST_COPY` as last parameter instead `MHD_RESPMEM_MUST_FREE` and as you suggested `o_free(response_buffer)` at the end of function and it works correctly.

Many thanks.

**Comment by babelouest** [comment]
Thanks for helping!
I'll close this bug when I'll patch the code.

**Comment by babelouest** [comment]
Hello @NazzarenoMarziale ,

I thought that the flag `MHD_RESPMEM_MUST_FREE` replaced by `MHD_RESPMEM_MUST_COPY` is not bad by itself, but may result on unwanted memory consumption, because the buffer is copied into memory, so free'd twice with the new flag.
What if I add a flag in the `struct _u_instance` called `int mhd_response_buffer_copy` that would be set to `false` by default, but if set to true, the flag will use `MHD_RESPMEM_MUST_COPY` instead of `MHD_RESPMEM_MUST_FREE`.

So the initialization would be like
```C
struct _u_instance instance;
ulfius_init_instance(&instance, 8080, NULL, NULL);
instance.mhd_response_buffer_copy = 1;
ulfius_start_framework(&instance);
// rest of the code
```

What do you think?

**Comment by NazzarenoMarziale** [comment]
I think that it is perfect. In this way who need this functionality can set this flag and use this new facility. Will be inserted inside 2.4 version? 

Many thanks.

**Comment by NazzarenoMarziale** [comment]
Another idea could be that if I set new memory functions using `o_set_alloc_funcs` this flag will be set to 1 automatically inside this function. Because if I change memory functions I have to set this flag, if it will be set automatically it will avoid possibile bugs. 
What do you think?

**Comment by babelouest** [comment]
I thought about setting automatically `mhd_response_buffer_copy`, not on `o_set_alloc_funcs` because this function belong to `orcania` and not `ulfius`, but in `ulfius_init_instance` for example.

But there were at least two problems with this approach:
- The compiler yelled at me when I tried to compare the address of `malloc` and the internal `struct o_mallot_t` variable, I didn't go further after this error message but I think it can be fixed though
- It may not be mandatory to change MHD flag for all cases and can decrease the performance or the memory consumption

So basically, to combine this bugfix without changing the performance or the behavior on existing applications, I prefer letting users change the flag `mhd_response_buffer_copy` by themselves and precisely document the new flag.

It's planned to be in the 2.4 release.

**Comment by NazzarenoMarziale** [comment]
Perfect! Thank you very much.

**Comment by laf0rge** [comment]
I am still observing this issue with libulfius-2.5.2, so the fix probably didn't make it into the 2.4.x branch.

I don't understand the intrinsics between ulfius and microhttpd, but if every library always free'd whatever it allocated, it wouldn't matter if microhttpd was using system malloc and ulfius used some other allocator.  So to hit this kind of bug, something allocated by ulfius appears to be free'd by microhttpd, which is something like a 'layering violation'.  As long as those allocations/frees are intertwined, o_set_alloc_funcs() is pretty much unusable for ulfius :(

**Comment by laf0rge** [comment]
I should have read more carefully.  After setting instance->mhd_response_copy_data = 1, the error is gone. Sorry for the noise!

**Comment by babelouest** [comment]
Ulfius is based on libmicrohttpd (MHD) for the http service part. Indeed they both manage their allocated resources on their own. For now libmicrohttpd doesn't allow to change memory management functions, so it's true that the scope of changing ulfius memory functions is limited.

The requirement for this change was for the calling app using Ulfius to free the response body by itself instead of asking Ulfius and therefore MHD to do it. In addition to the fact that the calling app may use a different `malloc/free` than Ulfius or MHD, it can also be possible that the calling app wants ro reuse the same response body without having to rebuild it on every call for example. But because of the backward compatibility, I couldn't make this setting by default (and it would have been overkill too).

**Comment by babelouest** [comment]
I answer to myself, when using `instance->mhd_response_copy_data = 1`, the calling app doesn't have to free `response->binary_body`, Ulfius will do it by itself. My bad.

**Comment by laf0rge** [comment]
> I answer to myself, when using `instance->mhd_response_copy_data = 1`, the calling app doesn't have to free `response->binary_body`, Ulfius will do it by itself. My bad.

thanks for confirming this. Also, it will do so with the proper memory allocator which was used by libulfius, not by the implicit assumption by libmicrohttpd that one can pass all dynamically allocated memory to free().  This way it keeps the symmetry: libulfius free's what it allocates, and libmicrohttpd frees what it allocates, but not vice-versa.

**Comment by babelouest** [comment]
I wrote a patch for libmicrohttpd to avoid using `instance->mhd_response_copy_data` in the future. In the next versions, ulfius will provide to MHD the `free` function to clean the response body. I will maintain the compatibility though so the option `instance->mhd_response_copy_data` will stay. But this patch will make everyone happy I think.

https://git.gnunet.org/libmicrohttpd.git/commit/?id=3d1b941137f9d8379e6e67d5abd57be5ae6ebe1a

**Comment by babelouest** [comment]
Update: libmicrohttpd already had a feature to achieve this since 0.9.61 it seems. I will use this existing feature instead of my patch, with a backward compatibility for systems with older MHD versions.

**Comment by laf0rge** [comment]
I actually think this bug is back, at least on Debian unstable with ulfius 2.3.7-1 and mhd 0.9.71-1 I'm again seeing crashes when using the custom memory allocator functions - despite using "instance->mhd_response_copy_data = 1".  The custom memory allocator (in this case talloc) crashes in a free call-back, as some memory is handed to it which was never allocated using the allocator call-back before.

It might be an idea to add a test case to ulfius using talloc similar to https://git.osmocom.org/osmo-cbc/tree/src/rest_api.c#n650 as talloc is quite good in asserting/aborting in case of any invalid operation, like free()ing something never allocated by it.

**Comment by laf0rge** [comment]
See also https://osmocom.org/issues/4932 for the related bug report in osmo-cbc, a project using ulfius + talloc.

**Comment by babelouest** [comment]
Hello @laf0rge ,

If you use Debian unstable, I assume you use ulfius 2.7.1, not 2.3.7.

About you issue, did you set the memory management functions before? That may be the problem.
https://github.com/babelouest/ulfius/blob/master/API.md#memory-management

If you did, can you open a new issue with a sample code? It may be a different bug from this one.
